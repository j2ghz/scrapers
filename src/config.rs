use std::path::PathBuf;

use serde::de::{self};
use serde::Deserialize;

#[derive(Deserialize, Debug, Clone)]
pub struct Config {
    pub scrapers: Vec<ScraperDefinition>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct ScraperDefinition {
    pub dest: PathBuf,
    pub urls: Vec<url::Url>,
    pub domain_whitelist: Vec<String>,
    pub steps: Vec<ScraperStep>,
}

#[derive(Deserialize, Debug, Clone)]
pub enum ScraperStep {
    ExtractHrefsFromHTML(HTMLSelector),
    DownloadImage(HTMLSelector),
}

#[derive(Debug, Clone, PartialEq)]
pub struct HTMLSelector(voyager::scraper::Selector);

impl AsRef<voyager::scraper::Selector> for HTMLSelector {
    fn as_ref(&self) -> &voyager::scraper::Selector {
        &self.0
    }
}

impl<'de> Deserialize<'de> for HTMLSelector {
    fn deserialize<D>(d: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let s = String::deserialize(d)?;
        let selector = voyager::scraper::Selector::parse(&s)
            .map_err(|e| de::Error::custom(format!("Parse error: {e:?}")))?;
        Ok(HTMLSelector(selector))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn parse_example() {
        let config_contents = std::fs::read("test_config.yaml").unwrap();
        let config = &serde_yaml::from_slice::<Config>(&config_contents).unwrap();
        insta::assert_debug_snapshot!(config);
    }
}
