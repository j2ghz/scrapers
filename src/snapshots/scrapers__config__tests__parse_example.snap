---
source: src/config.rs
expression: config
---
Config {
    scrapers: [
        ScraperDefinition {
            dest: "target/test_temp",
            urls: [
                Url {
                    scheme: "https",
                    cannot_be_a_base: false,
                    username: "",
                    password: None,
                    host: Some(
                        Domain(
                            "webscraper.io",
                        ),
                    ),
                    port: None,
                    path: "/test-sites/e-commerce/allinone",
                    query: None,
                    fragment: None,
                },
            ],
            domain_whitelist: [
                "webscraper.io",
            ],
            steps: [
                ExtractHrefsFromHTML(
                    HTMLSelector(
                        Selector {
                            selectors: [
                                Selector(.caption .title, specificity = 0x800),
                            ],
                        },
                    ),
                ),
                DownloadImage(
                    HTMLSelector(
                        Selector {
                            selectors: [
                                Selector(img.img-responsive, specificity = 0x401),
                            ],
                        },
                    ),
                ),
            ],
        },
    ],
}
